### PR TITLE
feat: add flag to listen on TCP

### PR DESCRIPTION
**Pull request by andrewrynhard**
Signed-off-by: Andrew Rynhard <andrew@rynhard.io>

### CODE REVIEW

File: cmd/runtime/main.go [hunk 1]
```diff
@@ -53,9 +56,28 @@ func run() error {
 		}
 	}()
 
-	l, err := net.Listen("unix", socketPath)
+	if err := os.Remove(socketPath); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			log.Fatal(err)
+		}
+	}
```
Comment (andrewrynhard) [inline cmd/runtime/main.go:63]: Fixing this, `git` is broken for me right now. Will push a fix in a bit.

File: cmd/runtime/main.go [hunk 2]
```diff
@@ -53,9 +56,28 @@ func run() error {
...
+		network = "unix"
+		address = socketPath
+
+		if err := os.Remove(socketPath); err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				log.Fatal(err)
+			}
+		}
+	} else {
+		network = "tcp"
```
Comment (invidian) [inline cmd/runtime/main.go:74]: Can't this be default? Then you can skip `else` block completely :)

File: cmd/runtime/main.go [hunk 3]
```diff
@@ -26,10 +27,12 @@ import (
 	"github.com/cosi-project/runtime/pkg/state/protobuf/server"
 )
 
+var addressAndPort string
```
Comment (invidian) [inline cmd/runtime/main.go:30]: Must it be a global variable?

File: cmd/runtime/main.go [hunk 4]
```diff
@@ -53,9 +56,28 @@ func run() error {
...
+		address string
+	)
+
+	if socketPath == "" {
+		network = "unix"
+		address = socketPath
+
+		if err := os.Remove(socketPath); err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				log.Fatal(err)
```
Comment (invidian) [inline cmd/runtime/main.go:70]: I think you should return error here instead of panicking.

**Review by invidian — COMMENTED** [review]
Some suggestions :)

File: cmd/runtime/main.go (hunk 4 above)
Comment (andrewrynhard) [reply cmd/runtime/main.go:70]: Whoops

File: cmd/runtime/main.go (hunk 3 above)
Comment (andrewrynhard) [reply cmd/runtime/main.go:30]: Was already the pattern used. I just went with it.

Comment (andrewrynhard) [reply cmd/runtime/main.go:30]: Happy to address in a follow up, unless we feel it should be addressed in this PR.

**Comment by andrewrynhard** [comment]
`make fmt` touched generated files. You can ignore.

File: api/v1alpha1/resource.pb.go [hunk 5]
```diff
@@ -9,11 +9,12 @@
 package v1alpha1
 
 import (
+	reflect "reflect"
+	sync "sync"
+
 	proto "github.com/golang/protobuf/proto"
 	protoreflect "google.golang.org/protobuf/reflect/protoreflect"
 	protoimpl "google.golang.org/protobuf/runtime/protoimpl"
-	reflect "reflect"
```
Comment (smira) [inline api/v1alpha1/resource.pb.go:15]: not sure why these changes popped up?

File: cmd/runtime/main.go [hunk 6]
```diff
@@ -26,10 +27,14 @@ import (
...
 
-var socketPath string
+var (
+	addressAndPort string
+	socketPath     string
+)
 
 func main() {
-	flag.StringVar(&socketPath, "socket-path", "/var/run/cosi-runtime.sock", "path to the UNIX socket to listen on")
+	flag.StringVar(&socketPath, "socket-path", "", "path to the UNIX socket to listen on")
```
Comment (smira) [inline cmd/runtime/main.go:36]: this changes default behavior by making runtime listen on TCP by default vs. file socket

I think having it listening on `*:50000` by default is not a good default

Comment (andrewrynhard) [reply cmd/runtime/main.go:36]: Default to unix socket? Default to `127.0.0.1`? Either way is fine by me.

Comment (smira) [reply cmd/runtime/main.go:36]: I would keep it as file socket, not saying it's more secure than localhost, but at least shows the intention better as the API doesn't have any auth at the moment.

File: cmd/runtime/main.go [hunk 7]
```diff
@@ -26,10 +27,14 @@ import (
...
 
-var socketPath string
+var (
+	addressAndPort string
+	socketPath     string
+)
 
 func main() {
-	flag.StringVar(&socketPath, "socket-path", "/var/run/cosi-runtime.sock", "path to the UNIX socket to listen on")
+	flag.StringVar(&socketPath, "socket-path", "/system/runtime.sock", "path to the UNIX socket to listen on")
```
Comment (andrewrynhard) [inline cmd/runtime/main.go:36]: Updated to align with the `engine`'s assumptions.

**Review by smira — APPROVED** [review]
